monero/wallet_client.go

```diff
@@ -221,49 +250,211 @@ func (c *walletClient) WaitForReceipt(req *WaitForReceiptRequest) (*wallet.Trans
+			return nil, fmt.Errorf("sweep failed waiting for receipt: %w", err)
+		}
+		log.Infof("Sweep transfer ID=%s of %s XMR (%s XMR fees) completed at height %d",
+			txID,
+			coins.FmtPiconeroAmtAsXMR(receipt.Amount),
+			coins.FmtPiconeroAmtAsXMR(receipt.Fee),
+			receipt.Height,
+		)
+		transfers = append(transfers, receipt)
+	}
+	return transfers, nil
+}
+
+func (c *walletClient) CreateABWalletConf(walletNamePrefix string) *WalletClientConf {
+	walletName := fmt.Sprintf("%s-%s", walletNamePrefix, time.Now().Format(common.TimeFmtNSecs))
+	walletPath := path.Join(path.Dir(c.conf.WalletFilePath), walletName)
+	conf := &WalletClientConf{
+		Env:                 c.conf.Env,
+		WalletFilePath:      walletPath,
+		WalletPassword:      c.conf.WalletPassword,
+		WalletPort:          0,
+		MonerodNodes:        c.conf.MonerodNodes,
+		MoneroWalletRPCPath: c.conf.MoneroWalletRPCPath,
+		LogPath:             c.conf.LogPath,
+	}
+	return conf
+}
+
+func createWalletFromKeys(
+	conf *WalletClientConf,
+	walletRestoreHeight uint64,
+	privateSpendKey *mcrypto.PrivateSpendKey, // nil for a view-only wallet
+	privateViewKey *mcrypto.PrivateViewKey,
+	address mcrypto.Address,
+) (WalletClient, error) {
+	if conf.WalletPort == 0 { // swap wallets need randomized ports, so we expect this to be zero
+		var err error
+		conf.WalletPort, err = getFreeTCPPort()
+		if err != nil {
+			return nil, err
+		}
+	}
+	// should be a one item list, we use the same node that the primary wallet is using
+	monerodNode := conf.MonerodNodes[0]
+
+	proc, err := createWalletRPCService(
+		conf.Env,
+		conf.MoneroWalletRPCPath,
+		conf.WalletPort,
+		path.Dir(conf.WalletFilePath),
+		conf.LogPath,
+		monerodNode,
+	)
+	if err != nil {
+		return nil, err
+	}
+	c := NewThinWalletClient(monerodNode.Host, monerodNode.Port, conf.WalletPort).(*walletClient)
+	c.rpcProcess = proc
+	c.conf = conf
+	err = c.generateFromKeys(
+		privateSpendKey, // nil for a view-only wallet
+		privateViewKey,
+		address,
+		walletRestoreHeight,
+		path.Base(conf.WalletFilePath),
+		conf.WalletPassword,
+	)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	acctResp, err := c.GetAddress(0)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	c.walletAddr = mcrypto.Address(acctResp.Address)
+	if c.walletAddr != address {
+		panic("addresses do not match")
+	}
+
+	bal, err := c.GetBalance(0)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+
+	log.Infof("Created wallet %s, balance is %s XMR (%d blocks to unlock), address is %s",
+		c.WalletName(),
+		coins.FmtPiconeroAmtAsXMR(bal.Balance),
+		bal.BlocksToUnlock,
+		c.PrimaryAddress(),
+	)
+	return c, err
 }
 
-// GenerateFromKeys creates a wallet from a given wallet address, view key, and optional spend key
-func (c *walletClient) GenerateFromKeys(
-	kp *mcrypto.PrivateKeyPair,
+// CreateSpendWalletFromKeys creates a new monero-wallet-rpc process and wallet from a given wallet address,
+// view key, and optional spend key
```

Comment:
is the spend key optional? shouldn't `CreateViewOnlyWalletFromKeys` be used in that case

Reply:
also does `*mcrypto.PrivateKeyPair` allow for nil spend keys?

Reply:
`CreateSpendWalletFromKeys` only takes a `PrivateKeyPair`. I don't think your design for `PrivateKeyPair` ever intended the type to allow the view key to be unset.  `NewPrivateKeyPairFromBytes` will throw an error, for example, if the view key is not set. We could document that and maybe panic if a nil value is passed to `NewPrivateKeyPair`. My intention was that package external users would only use `CreateSpendWalletFromKeys` for spend wallets and only use `CreateViewOnlyWalletFromKeys` for view-only wallets.

Reply:
okay, I guess update the comment then to say the spend key is required?